criu/cgroup.c

```diff
@@ -1700,6 +1793,17 @@ static int prepare_cgroup_sfd(CgroupEntry *ce)
 				pr_perror("\tCan't make controller dir %s", paux);
 				goto err;
 			}
+
+			ret = cgroup_mountinfo_contains(ctrl->cnames, ctrl->n_cnames, "xattr");
+			if (ret < 0) {
+				pr_err("Failed to check xattr in host mountinfo\n");
+				goto err;
+			}
+			if (ret) {
+				int ooff = strlen(opt);
+				snprintf(opt + ooff, sizeof(opt) - ooff, ",xattr");
+			}
+
 			if (mount("none", paux, "cgroup", 0, opt) < 0) {
```

Comment:
If we parse a host mount info, we can bind-mount a host cgroup mount instead of mounting a new one. In this case, we will not need to care about options. What do you think about this?

Reply:
If cgroup is mounted readonly on host we will get readonly bind-mount. These new mount was introduced initially to get writable cgroup mount.

Reply:
> If cgroup is mounted readonly on host we will get readonly bind-mount. These new mount was introduced initially to get writable cgroup mount.

These new mounts was introduced, because we ditn't want to search it. It was easier to mount it, instead of parsing mountinfo and etc.

> upd: More other we don't always have a mount on host, so we still would need two cases where we do both mount and bind-mount

Is this a problem? A cgroup mount can be in another namespace and we will not find a proper options for it here. Is this a problem?

Can we always mount cgroups with xattr?

Reply:
>>If cgroup is mounted readonly on host we will get readonly bind-mount. These new mount was introduced initially to get writable cgroup mount.
>
>These new mounts was introduced, because we ditn't want to search it. It was easier to mount it, instead of parsing mountinfo and etc.

But there is still a posibility of readonly mount on host, thus making separate cgroup mount looks more general solution for me then binding (which in general requires additional writable remount).

Actually, I was wrong about the initials. I found these patch:

```
commit dda742ec570a3f6fea6d88c9b82e22306bcb1422
Author: Andrew Vagin <avagin@openvz.org>
Date:   Wed Jul 16 17:46:00 2014 +0400

    cgroup: don't look up existing mount point

    A mount point, which is mounted by someone else, may be umounted in
    any moment.
```

So 1) If someone unmounts the host's-mount, the bind-mount may fail... 2) And even more my code will not see xattr in mountinfo and likely some warnings will be still printed. But (2) is not so important as warnings are a fair cost for inproper mounts setup on host.

>
>>upd: More other we don't always have a mount on host, so we still would need two cases where we do both mount and bind-mount
>
>>Is this a problem?

Either we chouse between mount/bind-mount or we chouse between xattr/noxattr I think it will be almost the same amount of code. We still need to parse mountinfo in any case.

>>A cgroup mount can be in another namespace and we will not find a proper options for it here. Is this a problem?

If we have no mount on host but it actually existed previousely and kernel still holds 'xattr' option somthere deep inside for us, we have no way to suppress the 'missmatch' warning when mounting.

```
[root@snorch criu]# echo "cgroup DEBUG" >  /dev/kmsg
[root@snorch criu]# cat /proc/self/mountinfo | grep systemd
29 27 0:26 / /sys/fs/cgroup/systemd rw,nosuid,nodev,noexec,relatime shared:6 - cgroup cgroup rw,xattr,name=system
[root@snorch criu]# umount /sys/fs/cgroup/systemd
[root@snorch criu]# mount -t cgroup -onone,name=systemd cgroup /sys/fs/cgroup/systemd
[root@snorch criu]# cat /proc/self/mountinfo | grep systemd
29 27 0:26 / /sys/fs/cgroup/systemd rw,relatime shared:6 - cgroup cgroup rw,xattr,name=systemd
[root@snorch criu]# dmesg | grep "DEBUG\|mount" | tail -n2
[616485.628903] cgroup DEBUG
[616541.852775] new mount options do not match the existing superblock, will be ignored
```

>
>Can we always mount cgroups with xattr?

No, these way we will have missmatches for all other cgroups which initially had no option.

To conclude: to avoid 1)fall back to normal mount when src is unmounted, 2)fall back to normal mount on bindmount fail on src racy umount, 3)additional remount of the mount to rw if src is ro - I believe it is better to leave the current approach.